[C++ API] Functional DataParallel

goldsborough opened the pull request:
This PR adds the functional version of `DataParallel` (i.e. `data_parallel`) to the C++ frontend.

For this, I had to:
1. Add "differentiable" versions of scatter and gather, which perform their inverse operation in the backward pass, to C++. I've added them under `torch/csrc/autograd/functions/comm.{h,cpp}`. I had to move some utilities from `VariableType.cpp` into `torch/csrc/autograd/functions/utils.h`, and changed them a bit to fix the `const_cast`s for which there were `TODO`s,
2. Implement the `replicate`, `parallel_apply` and the combining `data_parallel` functions in C++.

`replicate` is implemented based on our existing `clone()` interface, along with the ability to set the current device via `at::OptionsGuard` (so nice).

`parallel_apply` is implemented using `at::parallel_for` (CC @cpuhrsch) and [follows the code from PyTorch](https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/parallel_apply.py).

Added lots of tests for these things.

@apaszke @ezyang @ebetica @colesbury

ezyang commented:
Build failure looks legit.

goldsborough commented:
I'm trying to use the new `CUDAStream` interface we just got in ATen, but this is causing segfaults. Investigating.

apaszke commented on torch/csrc/api/include/torch/nn/parallel/data_parallel.h:
```diff
@@ -0,0 +1,161 @@
+    // thread local default options and thereby cause all tensors in the cloned
+    // module to be constructed directly on the device we want.
+    OptionsGuard guard(device);
+    replicas.push_back(std::static_pointer_cast<ModuleType>(module->clone()));
```
Hmm that's a bit surprising. I'd expect `clone()` to completely ignore the thread local device and just put them wherever the original params where. This way makes sense too I guess 😕

apaszke commented on torch/csrc/api/include/torch/nn/parallel/data_parallel.h:
```diff
@@ -0,0 +1,161 @@
+            std::cerr << "Replica #" << index
+                      << " terminated in parallel_apply() with exception: "
+                      << exception.what();
+            std::terminate();
```
That's not too elegant (both printing and terminating unconditionally). We stash the exception in Python land and reraise it later. I think it's possible even in C++ using `std::exception_ptr`

apaszke commented on torch/csrc/api/include/torch/nn/parallel/data_parallel.h:
```diff
@@ -0,0 +1,161 @@
+    const auto device_count = torch::cuda::device_count();
+    AT_CHECK(device_count > 0, "Expected at least one CUDA device");
+    devices.emplace();
+    devices->reserve(device_count);
```
Why do you reserve only once you emplace? This will likely trigger a reallocation. Also, aren't you constructing a vector where the first device appears twice?

goldsborough replied:
`emplace()` only creates the vector, `reserve` then allocates its memory. I think there are no unnecessary allocations?

apaszke replied:
Ok, I didn't notice it's an optional thing, so I understood that as `emplace_back()`.

apaszke commented on torch/csrc/api/include/torch/nn/parallel/data_parallel.h:
```diff
@@ -0,0 +1,161 @@
+  auto scattered_inputs = scatter.apply({input});
+
+  if (devices->size() == 1) {
+    return module->forward(input);
```
Can't you move this before the scatter?

apaszke commented on test/cpp/api/parallel.cpp:
```diff
@@ -0,0 +1,179 @@
+
+  auto input = torch::ones(10, torch::requires_grad(true));
+  auto output = scatter.apply({input});
+
```
Would be good to check that content of outputs is ok (move to a single device, use `cat`)

apaszke commented on test/cpp/api/parallel.cpp:
```diff
@@ -0,0 +1,179 @@
+
+  auto a = torch::ones(5, torch::requires_grad(true).device({torch::kCUDA, 0}));
+  auto b = torch::ones(5, torch::requires_grad(true).device({torch::kCUDA, 1}));
+  auto output = gather.apply({a, b});
```
Same as above

apaszke commented on test/cpp/api/parallel.cpp:
```diff
@@ -0,0 +1,179 @@
+
+  REQUIRE(b.grad().defined());
+  REQUIRE(b.grad().device() == torch::Device(torch::kCUDA, 1));
+  REQUIRE(b.grad().sum().toCInt() == 5);
```
We should have sth like `.eq(1).all()` in ATen, and it's a stronger check than this. Also, ideally you'd pass in a random output and check that the grads here are just slices

apaszke commented on test/cpp/api/parallel.cpp:
```diff
@@ -0,0 +1,179 @@
+TEST_CASE("Parallel/Replicate", "[cuda]") {
+  Linear linear(3, 4);
+  auto replicas = parallel::replicate(
+      linear, {torch::Device(torch::kCUDA, 0), torch::Device(torch::kCUDA, 1)});
```
Shouldn't we skip this test somehow if someone doesn't have 2 GPUs?

goldsborough replied:
Yes good point

apaszke commented on test/cpp/api/parallel.cpp:
```diff
@@ -0,0 +1,179 @@
+        replica2_parameters[i]->data().data<float>() !=
+        original_parameters[i]->data().data<float>());
+  }
+}
```
These two blocks look like an unrolled loop to me 😄

goldsborough replied:
I personally prefer to do this kind of duplication in test code as long as its not unreasonable, it makes debugging easier (e.g. the line information about where the assertion hit tells you more than if it was in a loop)

apaszke commented on test/cpp/api/parallel.cpp:
```diff
@@ -0,0 +1,179 @@
+      linear,
+      input,
+      /*devices=*/at::nullopt,
+      /*output_device=*/torch::Device(torch::kCUDA, 1));
```
Again, ignore if don't have 2 GPUs?

apaszke reviewed (commented):
Mostly LGTM. I have some comments that might help clean up the code. Would be good to fix the `std::terminate` in case of an exception before merging.

goldsborough commented on torch/csrc/autograd/functions/utils.h:
```diff
@@ -18,9 +21,59 @@ using function_constructor = std::function<std::shared_ptr<Function>(edge_list&&
+  for (auto& variable : variables) {
+    set_history(variable, grad_fn);
+  }
+}
```
It used to be a `const vector`, but we actually need to call `set_gradient_edge` on every variable. We used to have a `const_cast` here, but I've removed it for better style. Unfortunately we need these two overloads now. I could probably just make it a template though

goldsborough commented:
@apaszke I think I addressed all your comments now. Thanks for the review!

goldsborough committed:
Implement data parallel

Throw runtime error in non-CUDA environments

Remove THCStream forward declarations

Add retain() before converting from THCStream to CUDAStream

Conditionally compile with OpenMP in libtorch

Improve move-efficiency of comm.cpp and add multi-gpu guard

Fix single-device case of data_parallel

Include functional.h in python_comm.cpp

Rethrow exception in parallel_apply

Clarify data-parallel documentation

goldsborough committed:
Post rebase fixes

goldsborough commented:
@pytorchbot retest this please

goldsborough committed:
Remove unused stream_or_default

goldsborough commented:
@pytorchbot retest this please
